### PR TITLE
Extract MANIFEST.json from uploaded file

### PR DESCRIPTION
**Pull request by felipempda**
Small contribution considering #5
The idea is to use the file handle and tarfile package (that supports tar.gz files) to be able to extract the MANIFEST.json without contacting artifactory.
Thanks a lot for this project by the way!

### CODE REVIEW

**Comment by felipempda** [comment]
> I will need a little time to take a closer look at the code, and I may want to tweak some things a bit, is it ok if I push to your branch? Is the checkbox on this PR set so that I can do so?

Yes, I wouldn't mind that. "Allow edits by maintainers" is checked, that's the one right?
Be my guest, please. 

> On a side note, I'd like to know if the current method of extracting `MANIFEST.json` from Artifactory is a problem for your use of it (non-pro license? something else?). This is definitely something I want too, just curious your particular interest in it.

I was getting an unathorized error during uploading with ``` !MANIFEST.json```:

```shell
Traceback (most recent call last):
  File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1822, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1820, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1796, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/user/Projects//galactory/galactory/api/v2/collections.py", line 180, in publish
    manifest = load_manifest_from_artifactory(target)
  File "/home/user/Projects//galactory/galactory/utilities.py", line 63, in load_manifest_from_artifactory
    with urlopen(str(artifact) + '!/MANIFEST.json') as u:  
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unauthorized
```
Initially I was hoping artifactory would have native support for ansible galaxy collections but it didn't seem to work so that's when I saw your project. ;)

I was looking for something that would work as a proxy and local repository.
In other words the hability to deploy personnal collections in a internal repository but also capable of searching public online content in ansible.galaxy.com, etc. 

So I plan to set up galactory this way:
- GALACTORY_ARTIFACTORY_PATH = < internal arfiactory repo >
- GALACTORY_LISTEN_ADDR = "0.0.0.0"
- GALACTORY_LISTEN_PORT = "< port >"
- GALACTORY_NO_PROXY_NAMESPACE = "< personal namespace>"
- GALACTORY_PROXY_UPSTREAM = "https://galaxy.ansible.com"

So clients could have an unique local endpoint that would be able to provide collections regardless of where they are stored.

**Comment by briantist** [comment]
> > I will need a little time to take a closer look at the code, and I may want to tweak some things a bit, is it ok if I push to your branch? Is the checkbox on this PR set so that I can do so?
> 
> Yes, I wouldn't mind that. "Allow edits by maintainers" is checked, that's the one right? Be my guest, please.

Ok, so, I did pull down your branch and I started working on it; I really wanted to completely remove loading from Artifactory at all so we could close #5. After a little while, I realized that the changes I wanted to make were substantial and thought it better to put up a separate PR:
- #16 

I hope that's ok, I want to make it clear that I really appreciate this PR and I hope you'll be encouraged to contribute in the future too!

In the meantime, it'd be great if you can test that PR and see if it also works for you. And I'll keep this one open until that one's complete just in case.

> > On a side note, I'd like to know if the current method of extracting `MANIFEST.json` from Artifactory is a problem for your use of it (non-pro license? something else?). This is definitely something I want too, just curious your particular interest in it.
> 
> I was getting an unauthorized error when uploading during `!MANIFEST.json`:
> 
> ```shell
> Traceback (most recent call last):
>   File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 2525, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1822, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1820, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "/home/user/Projects//galactory/.env/lib/python3.8/site-packages/flask/app.py", line 1796, in dispatch_request
>     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
>   File "/home/user/Projects//galactory/galactory/api/v2/collections.py", line 180, in publish
>     manifest = load_manifest_from_artifactory(target)
>   File "/home/user/Projects//galactory/galactory/utilities.py", line 63, in load_manifest_from_artifactory
>     with urlopen(str(artifact) + '!/MANIFEST.json') as u:  
>   File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.8/urllib/request.py", line 531, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
>     response = self.parent.error(
>   File "/usr/lib/python3.8/urllib/request.py", line 569, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 401: Unauthorized
> ```

I see thanks! Are you using Artifactory OSS? Or is it a Pro License or higher?

> Initially I was hoping artifactory would have native support for ansible galaxy collections but it didn't seem to work so that's when I saw your project. ;)

Haha, yeah I was hoping the same, and I got tired of waiting for it. They have (had?) an open feature request but it's like 3+ years old and it seems like they have no intention of adding support for it.

> I was looking for something that would work as a proxy and local repository. In other words the hability to deploy personnal collections in a internal repository but also capable of searching public online content in ansible.galaxy.com, etc.

Perfect! Mind if I ask how you found it? Was it via [The Bullhorn](https://github.com/ansible/community/wiki/News#the-bullhorn)? I announce new releases there.

> So I plan to set up galactory this way:
> 
> * GALACTORY_ARTIFACTORY_PATH = < internal arfiactory repo >
> * GALACTORY_LISTEN_ADDR = "0.0.0.0"
> * GALACTORY_LISTEN_PORT = "< port >"
> * GALACTORY_NO_PROXY_NAMESPACE = "< personal namespace>"
> * GALACTORY_PROXY_UPSTREAM = "https://galaxy.ansible.com"
> 
> So clients could have an unique local endpoint that would be able to provide collections regardless of where they are stored.

Great! That's exactly the intention. So happy to hear you're finding it useful (or, trying to, hopefully once this upload issue is fixed).
